src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Workspace.cs

```diff
@@ -0,0 +1,510 @@
+        }
+
+        static string GetWorkspaceProjectContextId(string projectFilePath, Guid projectGuid, ProjectConfigurationSlice slice)
+        {
+            // WorkspaceContextId must be unique across the entire solution for the life of the solution, therefore as we fire
+            // up a workspace context per implicitly active config, we factor in both the full path of the project, the GUID of
+            // project and the name of the config. This will be unique across regardless of whether projects are added or renamed
+            // to match this project's original name. We include file path to make debugging easier on the Roslyn side.
+            //
+            // NOTE: Roslyn also uses this name as the default "AssemblyName" until we explicitly set it, so we need to make
+            // sure it doesn't contain any invalid path characters.
+            //
+            // For example:
+            //      C:\Project\Project.csproj ({72B509BD-C502-4707-ADFD-E2D43867CF45})
+            //      C:\Project\MultiTarget.csproj (net45 {72B509BD-C502-4707-ADFD-E2D43867CF45})
+            //      C:\Project\MultiTarget.csproj (net6.0 {72B509BD-C502-4707-ADFD-E2D43867CF45})
+
+            if (slice.Dimensions.Count == 0)
+                return $"{projectFilePath} ({projectGuid.ToString("B").ToUpperInvariant()})";
+            else
+                return $"{projectFilePath} ({string.Join(";", slice.Dimensions.Values)} {projectGuid.ToString("B").ToUpperInvariant()})";
+        }
+    }
+
+    private async Task OnBuildUpdateAsync(IProjectVersionedValue<BuildUpdate> update)
+    {
+        if (_initializationCompletion.Task.IsFaulted)
+        {
+            await _initializationCompletion.Task;
+        }
+
+        System.Diagnostics.Debug.Assert(_state is StateInitialized or StateInitializing);
+
+        Assumes.NotNull(_buildProgressRegistration);
+
+        await OnProjectChangedAsync(
+            _buildProgressRegistration,
+            update,
+            hasChange: static e => e.Value.BuildRuleUpdate.ProjectChanges[ProjectBuildRuleName].Difference.AnyChanges || e.Value.CommandLineArgumentsSnapshot.IsChanged,
+            applyFunc: ApplyProjectBuild,
+            _unloadCancellationToken);
+
+        _initializationCompletion.TrySetResult();
+
+        void ApplyProjectBuild(
+            IProjectVersionedValue<BuildUpdate> update,
+            ContextState state,
+            CancellationToken cancellationToken)
+        {
+            IProjectChangeDescription projectChange = update.Value.BuildRuleUpdate.ProjectChanges[ProjectBuildRuleName];
+
+            // There should always be some change to publish, as we have already called BeginBatch by this point
+            // TODO understand why the CLA snapshot's changed state differs from the project update, as they are supposed to travel together in sync
+            //Assumes.True(projectChange.Difference.AnyChanges && update.Value.CommandLineArgumentsSnapshot.IsChanged);
```

Comment:
nit: Commented code.

Reply:
The TODO relates to that. I plan to revisit the use of CommandLineArguments in a future PR that'll sort this out.

Reply:
The PR that'll remove the `CommandLineArgumentsSnapshot` is #8337.